subsys/power/device_pm.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <kernel.h>
+#include <device.h>
+#include <misc/__assert.h>
+
+#define LOG_LEVEL CONFIG_PM_LOG_LEVEL /* From power module Kconfig */
+#include <logging/log.h>
+LOG_MODULE_DECLARE(power);
+
+/* Device PM request type */
+#define DEVICE_PM_ASYNC			(1 << 0)
+
+static void device_pm_callback(struct device *dev,
+			       int retval, void *context, void *arg)
+{
+	u32_t target_state;
+
+	if (retval != 0) {
+		if (*((u32_t *)context) == DEVICE_PM_ACTIVE_STATE) {
+			target_state = DEVICE_PM_SUSPEND_STATE;
+		} else {
+			target_state = DEVICE_PM_ACTIVE_STATE;
+		}
+	} else {
+		target_state = *((u32_t *)context);
+	}
+
+	/* Lock only if callback is called asynchronously */
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
```

Comment:
I see one issue here with logic callback. 

device_pm_callback(void *context)
{
atomic_set(&fsm_state, (fsm_state_t)(context));
k_work_submit(fsm_work);
}

After callback when the worker gets called, the loop will again run forever because if we set the fsm to active/suspended in callback then again in the worker will run through the same logic.
we would need an additional variable (target state) to know the actual device state otherwise we would run through the loop continuously. 
this would mean we would need another atomic variable for target state?

Reply:
I realized we are using the usage count instead of target_state in the worker, but how do we handle device_set_power failure cases where the callback is called error code.

Reply:
@pizi-nordic Gentle reminder...

Reply:
> I see one issue here with logic callback.
> (...)
> After callback when the worker gets called, the loop will again run forever because if we set the fsm to active/suspended in callback then again in the worker will run through the same logic.

What loop? I do not see any loop in the proposed design.
- In the device_pm_callback you set the FSM state to either SUSPENDED to ACTIVE.
- In the work you are checking the FSM. If you are in SUSPENDED or ACTIVE state **and** only if the given state is not the desired one (obtained from the usage count), you are triggering next transition.

Reply:
I mentioned the loop w.r.t using target_state. If we use usage count we will not have loop the loop issue.

My concern was with error handling and with device_pm_enable/disable(). These are solved with use of target_state  atomic variable.

Reply:
>  I realized that we are using the usage count instead of target_state in the worker, but how do we handle device_set_power failure cases where the callback is called error code.

IMHO error here is a critical bug and we cannot continue. We can have 2 source of errors:
- Invalid state of the device (for example when we try suspending busy device): This is usage error. If the user use the device_get/put API correctly we will never run into this.
- I/O error (like I2C bus communication error or something similar): This is unrecoverable error at the moment. We cannot be sure what is real state of peripherar or even if we are still able to communicate with it. If there will be API to reset device, we would be able to use it to regain control over the device, but this will only delay the problem (what if reset fails?).

Reply:
So how do we handle the error case?

- we can either do ASSERT on error
- Or log the error and set the device to the previous state. If we set fsm state to the previous state it may not match the usage count, in this case, the loop will run forever if the callback return error again and again.